baybe/symmetries.py

```diff
@@ -0,0 +1,439 @@
+"""Functionality for expressing symmetries of the modeling process."""
+
+from __future__ import annotations
+
+import gc
+from abc import ABC, abstractmethod
+from collections.abc import Iterable
+from itertools import combinations
+from typing import TYPE_CHECKING, Any, cast
+
+import numpy as np
+import pandas as pd
+from attrs import Converter, define, field
+from attrs.validators import deep_iterable, ge, instance_of, min_len
+from typing_extensions import override
+
+from baybe.constraints.conditions import Condition
+from baybe.serialization import SerialMixin
+from baybe.utils.augmentation import (
+    df_apply_dependency_augmentation,
+    df_apply_mirror_augmentation,
+    df_apply_permutation_augmentation,
+)
+from baybe.utils.conversion import normalize_str_sequence
+from baybe.utils.validation import validate_is_finite, validate_unique_values
+
+if TYPE_CHECKING:
+    from baybe.parameters.base import Parameter
+    from baybe.searchspace import SearchSpace
+
+
+@define(frozen=True)
+class Symmetry(SerialMixin, ABC):
+    """Abstract base class for symmetries.
+
+    Symmetry is a concept that can be used to configure the modelling process in the
+    presence of invariances.
+    """
+
+    use_data_augmentation: bool = field(
+        default=True, validator=instance_of(bool), kw_only=True
+    )
+    """Flag indicating whether data augmentation would be used with surrogates that
+    support this."""
+
+    @property
+    @abstractmethod
+    def parameter_names(self) -> tuple[str, ...]:
+        """The names of the parameters affected by the symmetry."""
+
+    def summary(self) -> dict:
+        """Return a custom summarization of the symmetry."""
+        symmetry_dict = dict(
+            Type=self.__class__.__name__, Affected_Parameters=self.parameter_names
+        )
+        return symmetry_dict
+
+    @abstractmethod
+    def augment_measurements(
+        self, df: pd.DataFrame, parameters: Iterable[Parameter]
+    ) -> pd.DataFrame:
+        """Augment the given measurements according to the symmetry.
+
+        Args:
+            df: The dataframe containing the measurements to be augmented.
+            parameters: Parameter objects carrying additional information (might not be
+                needed by all augmentation implementations).
+
+        Returns:
+            The augmented dataframe including the original measurements.
+        """
+
+    def validate_searchspace_context(self, searchspace: SearchSpace) -> None:
+        """Validate that the symmetry is compatible with the given searchspace.
+
+        Args:
+            searchspace: The searchspace to validate against.
+
+        Raises:
+            ValueError: If the symmetry affects parameters not present in the
+                searchspace.
+        """
+        parameters_missing = set(self.parameter_names).difference(
+            searchspace.parameter_names
+        )
+        if parameters_missing:
+            raise ValueError(
```

Comment:
Is `IncompatibilityError` maybe more appropriate to raise in these validation methods (also in subclasses)? Because it indicates the the symmetry is `incompatible` with the search space, whereas `ValueError` would be more for things `within` the same class?